arch/risc-v/src/mpfs/mpfs_i2c.c

```diff
@@ -432,11 +432,24 @@ static int mpfs_i2c_sem_waitdone(struct mpfs_i2c_priv_s *priv)
 static int mpfs_i2c_irq(int cpuint, void *context, void *arg)
 {
   struct mpfs_i2c_priv_s *priv = (struct mpfs_i2c_priv_s *)arg;
-  struct i2c_msg_s *msg = &priv->msgv[priv->msgid];
+  struct i2c_msg_s *msg;
   volatile uint32_t status;
   uint8_t clear_irq = 1u;
 
-  DEBUGASSERT(msg != NULL);
+  /* On a very special case, i2c irq might arrive shortly after a warm
+   * reboot.  Try to detect that with some sanity checks.
+   */
+
+  if (priv == NULL)
```

Comment:
Why would the irq trigger after warm boot? Mss is reset, but what happens with the fpga? Maybe the fabric should be reset at warm boot as well?

Reply:
> Why would the irq trigger after warm boot? Mss is reset, but what happens with the fpga? Maybe the fabric should be reset at warm boot as well?

Still working with different build issues, as to verify whether this patch fixes any..

Reply:
Looks like it's an irq that comes very early for some reason. This patch makes it no longer crash, but will not ack the irq, so the system keeps flooding the irq giving no time for anything else; so the system appears stuck.

Looks like the ack is needed, and a few unnecessary checks may be avoided. Then it becomes stable; will do more testing.

Reply:
I suspect the corei2c irq line is asserted when we do the soft reset. When the interrupt source is enabled again, it will fire immediately and unexpectedly.

Is there a way to explicitly clear the interrupt on the corei2c device block ? Or explicitly reset the corei2c device ?

The way we do a soft reset is simply by:

`j __start`

So a reset signal is not emitted on the soc.

Reply:
Now the irq is enabled right before issuing the START condition, and disabled at the end of the transaction. Maybe I could try clearing the irq before enabling it, right.

Reply:
Clearing the The Serial Interrupt flag before enabling i2c or clearing it before starting any transaction doesn't help any. Not sure that unnecessary interrupt can be avoided at all.

Reply:
Might there be a FIFO in the i2c device that needs to be emptied too ? With SPI at least the RX interrupt didn't de-assert before the FIFO was also reset.

This is strange indeed, to me it would make sense that something does not reset properly and if we do the "j start" operation at a bad time (when the i2c transaction is running) we get old / stale interrupts from the device still.

Reply:
No FIFO, but data register. Could try read it out before enabling

Reply:
Most of the time it's: "Data byte has been received; NACK has been returned." or MPFS_I2C_ST_RX_DATA_NACK 0x58, so it maybe reboots between a transaction

Reply:
Also MPFS_I2C_ST_RX_DATA_ACK is present, I think if the system is 'rebooted' while in transaction, the peripheral device will continue the i2c transaction which eventually causes an unnecessary interrupt.